x-pack/elastic-agent/pkg/agent/application/emitter.go

```diff
@@ -30,54 +34,163 @@ type programsDispatcher interface {
-	return func(c *config.Config) error {
-		if err := InjectAgentConfig(c); err != nil {
-			return err
+type emitterController struct {
+	logger      *logger.Logger
+	controller  composable.Controller
+	router      programsDispatcher
+	modifiers   *configModifiers
+	reloadables []reloadable
+
+	// state
+	lock   sync.RWMutex
+	config *config.Config
+	ast    *transpiler.AST
+	vars   []transpiler.Vars
+}
+
+func (e *emitterController) Update(c *config.Config) error {
+	if err := InjectAgentConfig(c); err != nil {
+		return err
+	}
+
+	// perform and verify ast translation
+	m, err := c.ToMapStr()
+	if err != nil {
+		return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
+	}
+	rawAst, err := transpiler.NewAST(m)
+	if err != nil {
+		return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
+	}
+	for _, filter := range e.modifiers.Filters {
+		if err := filter(e.logger, rawAst); err != nil {
+			return errors.New(err, "failed to filter configuration", errors.TypeConfig)
 		}
+	}
 
-		log.Debug("Transforming configuration into a tree")
-		m, err := c.ToMapStr()
+	// sanitary check that nothing in the config is wrong when it comes to variable syntax
+	ast := rawAst.Clone()
+	inputs, ok := transpiler.Lookup(ast, "inputs")
+	if ok {
+		renderedInputs, err := renderInputs(inputs, []transpiler.Vars{
+			{
+				Mapping: map[string]interface{}{},
+			},
+		})
 		if err != nil {
-			return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
+			return err
 		}
-
-		ast, err := transpiler.NewAST(m)
+		err = transpiler.Insert(ast, renderedInputs, "inputs")
 		if err != nil {
-			return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
+			return err
 		}
+	}
 
-		for _, filter := range modifiers.Filters {
-			if err := filter(log, ast); err != nil {
-				return errors.New(err, "failed to filter configuration", errors.TypeConfig)
+	programsToRun, err := program.Programs(ast)
+	if err != nil {
+		return err
+	}
+
+	for _, decorator := range e.modifiers.Decorators {
+		for outputType, ptr := range programsToRun {
+			programsToRun[outputType], err = decorator(outputType, ast, ptr)
+			if err != nil {
+				return err
 			}
 		}
+	}
+
+	e.lock.Lock()
+	e.config = c
+	e.ast = rawAst
+	e.lock.Unlock()
+
+	return e.update()
+}
 
-		log.Debugf("Supported programs: %s", strings.Join(program.KnownProgramNames(), ", "))
-		log.Debug("Converting single configuration into specific programs configuration")
+func (e *emitterController) Set(vars []transpiler.Vars) {
+	e.lock.Lock()
+	ast := e.ast
+	e.vars = vars
+	e.lock.Unlock()
 
-		programsToRun, err := program.Programs(ast)
+	if ast != nil {
+		err := e.update()
 		if err != nil {
-			return err
+			e.logger.Errorf("Failed to render configuration with latest context from composable controller: %s", err)
```

Comment:
Would it be an option to return the error instead of logging it here? This could make testing easier.

Reply:
It is actually very un-likely that an error will actually occur here, because most errors would come from invalid configuration files, like bad variable syntax or something of that nature. That is why the `Update()` returns the error, because that is caused when the configuration is parsed and will return an error back to the user or back to Fleet.

In the case that a provider changes a value in a variable or a context is removed it will just cause inputs to be add/update/removed. This can also happen at any time, so really logging it here is the only thing we can do.